src/word_debt_bot/cogs/cmd_err_handler.py

```diff
@@ -0,0 +1,66 @@
+from discord.ext import commands
+
+import word_debt_bot.client as client
+import word_debt_bot.game as game
+
+
+class CmdErrHandler(commands.Cog, name="Command Error Handler"):
+    def __init__(self, bot: client.WordDebtBot, game: game.WordDebtGame):
+        self.bot = bot
+        self.game = game
+
+    @commands.Cog.listener()
+    async def on_command_error(self, ctx, err: commands.CommandError):
+        """The event triggered when an error is raised while invoking a command.
+        Parameters
+        ------------
+        ctx: commands.Context
+            The context used for command invocation.
+        error: commands.CommandError
+            The Exception raised.
+        """
+
+        # This prevents any commands with local handlers being handled here in on_command_error.
+        if hasattr(ctx.command, "on_error"):
+            return
+
+        if isinstance(err, commands.CommandNotFound):
+            await ctx.send(
+                "Unknown command! Maybe ask Toast to make it?\n"
+                "For a list of available commands type '.help'."
+            )
+
+        elif isinstance(err, commands.BadArgument):
+            await ctx.send(
+                f"Invalid inputs were supplied for {ctx.command}!\n"
+                f"For more information type '.help {ctx.command}'."
+            )
+
+        elif isinstance(err, commands.MissingRequiredArgument):
+            await ctx.send(
+                f"Not all required inputs were given for {ctx.command}!\n"
+                f"For more information type '.help {ctx.command}'."
+            )
+
+        elif isinstance(err.__cause__, AttributeError) and not self.game:
+            await ctx.send("Game not loaded. (Yell at Toast!)")
+
+        elif isinstance(err, commands.CommandInvokeError):
```

Comment:
Is this error one that wraps around all exceptions thrown by commands internally? Wondering if we can't also throw in handling for `ValueError` (generally because arguments were invalid) and `KeyError` (often if a user isn't registered).

Reply:
From my understanding of `CommandInvokeError` that is correct. "Exception raised when the command being invoked raised an exception." Has attribute `original` that can be retrieved with `__cause__` which is an `Exception`.

Removing the `try` `except` from the `log` command (like below) is then handled by the `CommandInvokeError` in `cmd_err_handler.py`.

```python
    @commands.command(name="log")
    async def log(self, ctx, words: int):
        new_debt = self.game.submit_words(str(ctx.author.id), words)
        self.journal({"command": "log", "words": words, "user": str(ctx.authorid)})
        await ctx.send(f"Logged {words:,} words! New debt: {new_debt:,}")
```

So I'm confident we could migrate the `try` `except` blocks in `game_commands.py` if we wish, and even go so far as basing the `if` logic off of `ctx.command` as well for more accurate error messages. I like this idea too because it moves error handling all to one area rather than splitting between two files and cleans up the game command methods a little bit.

-> Added to TODOs

Reply:
#### Update:
1. Migrated the `try` `except` blocks from `game_commands.py` to the error handler.
2. Added remaining 'expected' error types to handler.
3. Updated unit tests related to error handling.

#### Further Actions:
The unit tests involving the `log` command are currently only ensuring that the command is raising the exception specified. I have yet to find a replacement for the `invoke()` method in these situations since it cannot take arguments other than `ctx`, I've posted a question on [StackOverflow](https://stackoverflow.com/questions/77883736/how-to-ensure-discord-bot-is-sending-message-from-on-command-error-when-a-comm) see if anything turns up.

However, the bot currently responds appropriately to such errors for example:
![Screenshot 2024-01-27 101110](https://github.com/Swiddis/word-debt-bot/assets/134148713/499e28c8-8b68-4c90-84b9-32e67d14b65f)

Reply:
#### Update:

Tests are working now, I wrapped them in another command to invoke to pass in the parameters. 

Now investigating some pytest warnings for `RuntimeWarning: coroutine 'AsyncMockMixin._execute_mock_call' was never awaited`

```
tests/test_game_commands_cog.py::test_submit_words_with_error
tests/test_game_commands_cog.py::test_submit_words_with_no_register
tests/test_game_commands_cog.py::test_submit_words_no_game
tests/test_game_commands_cog.py::test_request_leaderboard_no_game
  C:\Users\gregn\AppData\Local\pypoetry\Cache\virtualenvs\word-debt-bot-yFc_9xiU-py3.11\Lib\site-packages\discord\ext\commands\core.py:656: RuntimeWarning: coroutine 'AsyncMockMixin._execute_mock_call' was never awaited
```